### ISSUE TITLE
After enabling Engintron all my visitors IP are my VPS IP - New VPS - Tried solutions on #69 and #489

### ISSUE THREAD

**Issue by brunocerq** [engintron/engintron#1178]
HI,
I know this is a well know issue, but I have read several articles and solutions but can't seem to correct the issue.

Running the latest versions of all software, CENTOS 7.7 kvm [srv] v86.0.18 with all software updates.
Engintron v1.13.0 Build 20200328 (Nginx version: 1.16.1)

I'm using cloudflare and I have enabled the custom rule for it.

set $PROXY_DOMAIN_OR_IP "xxx.xxx.xxx.xxx"; <- placed my VPS public IP here.

I can access my websites, and when running the test PHP file that I grabbed from here (https://engintron.com/docs/#/pages/Check-if-visitor-IPs-are-properly-logged-in-Apache-after-installing-Engintron):

I have mod_remoteip installed, I have changed the h for an a in the log format of apache.

In the page that runs the PHP snipet I can see my real IP somewhere down there:

**Are Visitor IPs Getting Properly Logged On My Server?**

Your server's IP address is | Your IP address (as logged in the server) is
-- | --
xxxxxxx | xxxxxx <- THOSE are the SAME and are my VPS public IP

Verdict: **NO**

    [REMOTE_ADDR] => xxx.xxx.xxx.xxx <- VPS public server ip
    [SERVER_PORT] => 443
    [SERVER_SOFTWARE] => Apache
    [HTTP_CDN_LOOP] => cloudflare
    [HTTP_CF_CONNECTING_IP] => xxxxxxx <- MY REAL IP IS HERE
    [HTTP_X_REAL_IP] => xxxxxxx <- MY REAL IP IS HERE
    [HTTP_X_FORWARDED_PORT] => 443
    [HTTP_X_FORWARDED_FOR] => xxxxxx my real ip, xxxxx my real ip

So, I don't know what can be wrong.
My websites see all visitors coming from my VPS server IP and not by their real ip.

**Comment by brunocerq** [comment]
Threads I watched and solutions tried:
https://github.com/engintron/engintron/issues/69
https://github.com/engintron/engintron/issues/489

**What I find out is that I don't have this file:
/usr/local/apache/conf/includes/remoteip.conf**

I tried: 
- went to apache configuration and disabled the mod_remoteip, than went to engintron and clicked on "Restore Nginx IP forwarding in Apache"
It installed the mod_remoteip and from the installation logs all seams ok, but I still don't have the remoteip.conf file.

- The httpd.conf has the %h replaced by %a - didn't work
- Tried replacing the %h for %{X-Real-IP}i - didn't work
- Tried the alternative solution replacing %h %l" to "%h %a %l (note that the %h is still there but the %a added next) - didn't work.

(tested always with a rebuild and restart)

Clients appear always with my server IP and not their own.

**Comment by brunocerq** [comment]
I'm running a VPS on Google Cloud Engine, It has a Private Internal IP, and a Public static external IP.
Done some tests I saw **fevangelou** asking to another person:

ip addr show | grep 'inet ' | grep ' brd ' | cut -d/ -f1 | cut -c10- | tr '\n' ' '
Resolves to my **internal** private IP (apache logs all clients with my **External** IP, and **not their own**)

ifconfig
eth0: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1460
        inet **XXXXXXXX (internal private ip)**  netmask 255.255.255.255  broadcast xxxxxxxx
        
lo: flags=73<UP,LOOPBACK,RUNNING>  mtu 65536
        inet 127.0.0.1  netmask 255.0.0.0
        inet6 ::1  prefixlen 128  scopeid 0x10<host>
        
My public external IP does not appear here.

Don't know if this info helps to solve the issue.

**Comment by brunocerq** [comment]
Just to summarise the info I get from the ip.php test because on my first message there was an item missing, I'll put in here the relevant ones only:   

    [REMOTE_ADDR] => Resolves my server Public IP
    [SERVER_PORT] => 443
    [SERVER_ADDR] => Resolves my server internal Private IP
    [SERVER_NAME] => Resolves the domain from witch I accessed this page
    [SERVER_SOFTWARE] => Apache
    [HTTP_CDN_LOOP] => cloudflare
    [HTTP_CF_CONNECTING_IP] => Resolves correctly my Client IP
    [HTTP_X_REAL_IP] => Resolves correctly my Client IP
    [HTTP_X_FORWARDED_SERVER] => Resolves correctly the domain from where this file is
    [HTTP_X_FORWARDED_PROTO] => https
    [HTTP_X_FORWARDED_PORT] => 443
    [HTTP_X_FORWARDED_HOST] => Resolves correctly the domain from where this file is
    [HTTP_X_FORWARDED_FOR] => IP, IP => Resolves correctly my Client IP (2x)

The only one that even after I have sent the correct log formats still only sees my public IP and not my visitors IP is Apache... don't know what else to do.

Also don't know why I don't have an /usr/local/apache/conf/includes/remoteip.conf file.

**Comment by fevangelou** [comment]
This is not a "known issue" really. You use CloudFlare and as such the only way to pass the real visitor IP is to adapt log formatting to use `%{CF-Connecting-IP}i` instead.

**Comment by brunocerq** [comment]
Tried that, didn't work:

<IfModule mod_log_config.c>
    LogFormat "%v:%p %{CF-Connecting-IP}i %l %u %t \"%r\" %>s %b \"%{Referer}i\" \"%{User-Agent}i\"" combinedvhost
    <IfModule logio_module>
        LogFormat "%v %{%s}t %I .\n%v %{%s}t %O ." bytesvhost
    </IfModule>
    LogFormat "%v:%p %{CF-Connecting-IP}i %l %u %t \"%r\" %>s %b \"%{Referer}i\" \"%{User-Agent}i\"" combined
    LogFormat "%v:%p %{CF-Connecting-IP}i %l %u %t \"%r\" %>s %b" common
    LogFormat "%{Referer}i -> %U" referer
    LogFormat "%{User-agent}i" agent
    <IfModule logio_module>
        CustomLog "|/usr/local/cpanel/bin/splitlogs --dir=/etc/apache2/logs/domlogs --main=srv.bmc-hosting.com --s$
    </IfModule>
    CustomLog "|/usr/local/cpanel/bin/splitlogs --dir=/etc/apache2/logs/domlogs --main=srv.xxxxxx.com --maino$
</IfModule>

**Comment by fevangelou** [comment]
I'm not fluent in Apache's log options but you may need to put `%{CF-Connecting-IP}i` in escaped double quotes. Additionally, the order of variables matters to Apache. So `%v:%p` in the beginning of each line is wrong. Stick to the defaults as mentioned in that Engintron docs link.

**Comment by brunocerq** [comment]
The %v:%p is placed there automatically by apache configuration when you use the Apache configuration on WHM since there is now a field to update log formats there, anyways I edit the file directly and removed the %v:%p but it's the same.
The double quotes don't work since they are already inside a double quote statement and adding them will result in an error that prevents apache to restart.

Any additional ideas? things to try?

**Comment by fevangelou** [comment]
I believe this is what you need: https://support.cloudflare.com/hc/en-us/articles/360029696071

Follow it to the letter and let me know.

**Comment by brunocerq** [comment]
well that cloudflare page set an example for debian... in centos things are very different, but with some digging I managed to sort it out, since apache in centos places some files in different locations, so here is the solution for anyone with the same problem:

1 - Go to EA4 -> Apache Modules and make sure **mod_remoteip** is installed (mine was)
2 - Go to mod_remoteip conf file, in centos this file is here :  /etc/apache2/conf.modules.d/370_mod_remoteip.conf
3 - edit the file with this:
LoadModule remoteip_module modules/mod_remoteip.so
RemoteIPHeader        **X-Forwarded-For**
RemoteIPInternalProxy 127.0.0.1 **IP1 IP2**
RemoteIPHeader CF-Connecting-IP
**RemoteIPTrustedProxy 173.245.48.0/20
RemoteIPTrustedProxy 103.21.244.0/22
RemoteIPTrustedProxy 103.22.200.0/22
RemoteIPTrustedProxy 103.31.4.0/22
RemoteIPTrustedProxy 141.101.64.0/18
RemoteIPTrustedProxy 108.162.192.0/18
RemoteIPTrustedProxy 190.93.240.0/20
RemoteIPTrustedProxy 188.114.96.0/20
RemoteIPTrustedProxy 197.234.240.0/22
RemoteIPTrustedProxy 198.41.128.0/17
RemoteIPTrustedProxy 162.158.0.0/15
RemoteIPTrustedProxy 104.16.0.0/12
RemoteIPTrustedProxy 172.64.0.0/13
RemoteIPTrustedProxy 131.0.72.0/22
RemoteIPTrustedProxy 2400:cb00::/32
RemoteIPTrustedProxy 2606:4700::/32
RemoteIPTrustedProxy 2803:f800::/32
RemoteIPTrustedProxy 2405:b500::/32
RemoteIPTrustedProxy 2405:8100::/32
RemoteIPTrustedProxy 2a06:98c0::/29
RemoteIPTrustedProxy 2c0f:f248::/32**

Note: in **RemoteIPInternalProxy** example I placed up there I only had initially one IP, so it was like this: 127.0.0.1 myInternalPrivateIP
Since I have 2 IP's (by the way I'm on a Google Cloud Compute Engine VPS) I had to **ADD** also my Public IP so in the end that line now has 127.0.0.1 **myInternalPrivateIP** **myPublicExternalIP**

Save, restart apache... and it's working.
By the way, in logging, I used the %a replacing the %h, with this configuration there is no need for the %{CF-Connecting-IP}i or %{X-Real-IP}i.

Hope this helps someone with the same setup I have to easily fix this issue.
